### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by lorehov** [lorehov/django_context_logging#5]
@bersace 
I think we should discuss roadmap.

The main thing that bothers me is that initially I called this package django_context_logging. If we want to add more generic logic or celery related logic - then may be better to rename a package?

**Comment by bersace** [comment]
Good idea. :-)

I wanted to open source this code and started by searching if anyone have already a project matching our requirement. I just found your project, studied it and feel it is the right project for this.

Last year i wrote a bunch of code to manage a transparent log context to send rich log records to log services (aka, graylog, sentry, logmatic, loggly, etc.).

We needed to have several shared fields in log records : app name, request id, task id (celery), customer id, user id, and some business related fields like document id, etc. We share this fields with other softwares in python, java and javascript (ember app).

The use cases for the developer, technical support, etc. are :

- a 500 error is reported by sentry, the developer wants to track all the logs of the corresponding requests in graylog (or wathever)
- a celery task failed, the dev wants to track all the logs of the task execution in gray log
- a customer hit a bug, technical support track history of a specific user, document, etc.
- when using logmatic or loggly, we tag each log with the source environment (staging, prod) to narrow the scope of search to one env, while using a single index in the service (for money or for cross-reference).

So the needs are :

 - have a thread-local context fed / unfed during the life of a requests / task
 - inject fields from settings (static)
 - inject / remove fields from middleware (requests, task). We have two request middleware:
   1. outermiddleware (the first in the onion) with associate the request id and clear it
   1. innermiddleware, push current customer/tenant and logged in user in the context (auth middleware has run)
 - inject / remove fields from the code at any time (`with` context manager is cool for this)
 - associate extrator to objects, and let the code just push an object *in the context*. e.g. : `log_context.push(request.user, requestId=request.id)`.
 - maybe other cool features :-)

All of this is about a thread local logging context with the nice filter :-) This is why i thought i should contribute to your project rather than creating a new one.

What do you think of this vision ? :-)

**Comment by lorehov** [comment]
My cases and needs is similar to yours (Thought we don't use SaaS logging solutions and currently we have no celery (what a shame :) )). But seems like your cases are more broad and you have written much more code. 
If it's true I suggest you to create new repo with name like `context_logging`. I also suggest to design it like follow: 
```
context_logging/
  |--ext/
     |--django/  <-- I could make PR with my django stuff here :)
     |--celery/
  |--context_logging.py
```
If we do not need to bring any non django related stuff then we could keep commit to this project.

**Comment by bersace** [comment]
So you suggest to keep a minimal libs for simple projects like yours and a full-featured lib for more complex projects ?

**Comment by bersace** [comment]
`pylogctx` would be great for a name :-)

**Comment by lorehov** [comment]
Not exactly. I think that project name should correctly describe what a project does.
So if it is a django_context_logging then this is a bunch of helpers for context logging in django. And if it is a `pylogctx` then there are could be many cool things (because name is much more generic).
Why? Because if I'm looking for a context logger for celery I will not event look at package with name `django_context_logging` :)

P.S.: My project is far from simple :) It is big, twisted and hard as hell (rasp.yandex.com) :-D We even have map reduce here.

**Comment by bersace** [comment]
Ok, let's fork then. :-)

**Comment by lorehov** [comment]
And if you start `pylogctx` and accept my pull request with django staff - then I switch to your library in my project.

**Comment by bersace** [comment]
I just created the fork and added you as admin. You're welcome. Thanks for the discussion and collaboration up to now :-)

**Comment by bersace** [comment]
@lorehov https://github.com/novafloss/pylogctx is in pretty good shape. What do you think of it ?

**Comment by lorehov** [comment]
@bersace agreed. In my projects I'd already switch to it. So I think I'll remove django_context_logging in order to not confuse people :)
